controller/contour/deployment.go

```diff
@@ -0,0 +1,334 @@
+	deployPodEnvVar = "POD_NAME"
+	// certsVolumeName is the name of the contour certificates volume.
+	certsVolumeName = "contourcert"
+	// certsVolumeMountPath is the mount path name of the contour certificates volume.
+	certsVolumeMountPath = "/certs"
+	// certsSecretName is the name of the secret used as the certificate volume source.
+	certsSecretName = certsVolumeName
+	// cfgVolumeName is the name of the contour configuration volume.
+	cfgVolumeName = "contour-config"
+	// cfgVolumeMountPath is the mount path name of the contour configuration volume.
+	cfgVolumeMountPath = "/config"
+)
+
+// ensureDeployment ensures a Deployment exists for the given contour.
+func (r *Reconciler) ensureDeployment(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+	haveDepl, current, err := r.currentDeployment(ctx, contour)
+	if err != nil {
+		return err
+	}
+
+	desired, err := desiredDeployment(contour)
+	if err != nil {
+		return fmt.Errorf("failed to build deployment: %v", err)
+	}
+
+	switch {
+	case !haveDepl:
+		if err := r.createDeployment(ctx, desired); err != nil {
+			return err
+		}
+	case haveDepl:
+		if _, err := r.updateDeployment(ctx, current, desired); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ensureContourDeleted ensures the Deployment resource for the provided
+// contour is deleted.
+func (r *Reconciler) ensureContourDeleted(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+	key := types.NamespacedName{
+		Namespace: contour.Spec.Namespace.Name,
+		Name:      contour.Name,
+	}
+
+	deployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: key.Namespace,
+			Name:      key.Name,
+		},
+	}
+
+	if err := r.Client.Delete(ctx, deployment); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	}
+
+	r.Log.Info("deleted deployment", "namespace", deployment.Namespace, "name", deployment.Name)
+	return nil
+}
+
+// desiredDeployment returns the desired deployment for the provided contour.
+func desiredDeployment(contour *operatorv1alpha1.Contour) (*appsv1.Deployment, error) {
+	label := map[string]string{"app": contour.Name}
+	key := types.NamespacedName{
+		Namespace: contour.Spec.Namespace.Name,
+		Name:      contour.Name,
+	}
+
+	deploy := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: key.Namespace,
+			Name:      key.Name,
+			// associate the deploy with the provided contour.
+			Labels: label,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &contour.Spec.Replicas,
+			// Ensure the deploy adopts only its own pods.
+			Selector: deploymentPodSelector(contour),
+		},
+	}
+
+	// TODO: Uncomment when https://github.com/projectcontour/contour-operator/pull/30 merges.
+	//deploy.SetOwnerReferences([]metav1.OwnerReference{newContourOwnerRef(contour)})
+
+	deploy.Spec.Template.Labels = deploymentPodSelector(contour).MatchLabels
+
+	annotations := map[string]string{"prometheus.io/scrape": "true", "prometheus.io/port": "8000"}
```

Comment:
A different TODO, but we need to update Contour to show how to use the Prometheus Operator to scrape Contour/Envoy metrics. Doing that then wouldn't require this annotation.

Reply:
@stevesloka do you have a contour/issue to xref?